src/qt/pivx.cpp

```diff
@@ -282,10 +277,11 @@ void BitcoinCore::initialize()
     }
 }
 
-void BitcoinCore::restart(QStringList args)
+void BitcoinCore::restart(const QStringList& args)
 {
-    if (execute_restart) { // Only restart 1x, no matter how often a user clicks on a restart-button
-        execute_restart = false;
+    static std::atomic<bool> restartAvailable{true};
+    static bool restart = true;
+    if (restartAvailable.compare_exchange_strong(restart, false)) {
```

Comment:
This kind of pattern does not work as intended.
With `compare_exchange_strong (T& expected, T val)`, if the comparison is true, the contained value is replaced with `val`, but, if the comparison is false, then `expected` is *overwritten* with the contained value.

In our case `restart` is first true, but, after comparing it with `restartAvailable=false` it becomes false as well, and at that point the comparison is always true. Further, the operation is atomic for `restartAvailable`, but not for `restart`.

This can be verified quickly with a small test like
```cpp
std::atomic<bool> ready (false);

void testfun(int id)
{
    std::string yeah = strprintf("Inside %d\n", id);
    std::string nope = strprintf("Nope %d\n", id);
    while (!ready) {}

    static std::atomic<bool> check1{true};
    static bool check2 = true;
    if (check1.compare_exchange_strong(check2, false)) {
        std::cout << yeah;
    } else {
        std::cout << nope;
    }
}

BOOST_AUTO_TEST_CASE(test_compare_exchange_strong)
{
    std::vector<std::thread> threads;
    for (int i=1; i<=20; ++i) threads.push_back(std::thread(testfun,i));
    ready = true;
    for (auto& th : threads) th.join();
}
```
we enter the loop for almost all threads.

I think that we can achieve our goal without this ping-pong, simply doing:
```cpp
static std::atomic<bool> restartAvailable{true};
if (restartAvailable.exchange(false)) {
    ...
}
``` 

`atomic::exchange` will read-and-replace atomically the value and return the value that was previously stored, which is exactly what we need.

Reply:
simpler, great 👍 .
I missed the atomic exchange function.